### PR TITLE
feat: Save nightly build in github releases with date tags

### PR DESCRIPTION
**Pull request by dmvict**
# Description

A solution to save nightly releases with date tag

## Problem\*

From issue:
> Similar to Rust, it would be great to be able to do something like noirup -v nightly-2023-08-3
Our current workflow is that each night, we overwrite the nightly tag, so there is no convenient way to get the nightly release from two or three days ago.
This is mainly a problem for project whom don't want to build from source, and don't want to wait for a release but they want to use the latest features.

Resolves https://github.com/noir-lang/noir/issues/2332

## Summary\*

- Added step to [get formatted date string](https://github.com/noir-lang/noir/compare/master...dmvict:save_nightly_releases?expand=1#diff-551d1fcf87f78cc3bc18a7b332a4dc5d8773a512062df881c5aba28a6f5c48d7R140)
- Updated release step to [generate tags/releases in format `nightly-yyyy-mm-dd`](https://github.com/noir-lang/noir/compare/master...dmvict:save_nightly_releases?expand=1#diff-551d1fcf87f78cc3bc18a7b332a4dc5d8773a512062df881c5aba28a6f5c48d7R154)

## Documentation

- This PR requires no documentation updates when merged.

## Additional Context

The step `Get formatted date` can be moved to job [`build-barretenberg`](https://github.com/noir-lang/noir/blob/master/.github/workflows/publish.yml#L25) and its output can be saved in job outputs. It reduces number of duplicates but the move will mix job purpose. Also we can create a separate job for the step. 

# PR Checklist\*

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by phated — DISMISSED** [review]
This breaks noirup since it doesn't keep our rolling nightly

**Comment by kevaundray** [comment]
Extending Blaine's comment; how does it solve this part of the issue:

>Each night when CI makes a release, it should tag the release artifacts with the date. **It should also tag the latest release artifact with nightly** so users can do noirup -v nightly to get the latest nightly version.

_Emphasis mine_

?

**Comment by dmvict** [comment]
> Extending Blaine's comment; how does it solve this part of the issue:
> 
> > Each night when CI makes a release, it should tag the release artifacts with the date. **It should also tag the latest release artifact with nightly** so users can do noirup -v nightly to get the latest nightly version.
> 
> _Emphasis mine_
> 
> ?

Thank you @phated  and @kevaundray 

Yes, the update was incorrect. I fixed it in last commit. Please, review the new changes.

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -148,6 +148,22 @@ jobs:
...
+        if: ${{ inputs.tag == '' && inputs.publish || github.event_name == 'schedule' }}
+        run: echo "date=$(date '+%Y-%m-%d')" >> $GITHUB_OUTPUT
+
+      - name: Upload binaries to nightly release with date tag
+        uses: svenstaro/upload-release-action@v2
+        if: ${{ inputs.tag == '' && inputs.publish || github.event_name == 'schedule' }}
+        with:
+          repo_name: noir-lang/noir
+          repo_token: ${{ secrets.GITHUB_TOKEN }}
+          file: ./nargo-${{ matrix.target }}.zip
```
Comment (kobyhallx) [inline .github/workflows/publish.yml:162]: Could we retain tar.gz extension?

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -148,6 +148,22 @@ jobs:
...
+          asset_name: nargo-${{ matrix.target }}.zip
+          overwrite: true
+          tag: ${{ format('{0}-{1}', 'nightly', steps.date.outputs.date) }}
```
Comment (kobyhallx) [inline .github/workflows/publish.yml:165]: I think this would no longer need nightly since we have date.

Comment (phated) [reply .github/workflows/publish.yml:165]: I think we should call these "previews" maybe, so it'd be `preview-12-12-23` or whatever. I'm also fine keeping "nightly" but I want something in the tag to indicate these are not official releases

Comment (kevaundray) [reply .github/workflows/publish.yml:165]: Yeah something in the tag to indicate that its nightly makes sense. I think preview is clearer, though "nightly build" keeps us inline with Rust and is a common idiom, so would prefer to keep that terminology

File: .github/workflows/publish.yml (hunk 1 above)
Comment (dmvict) [reply .github/workflows/publish.yml:162]: Hello @kobyhallx 

Should we save only the `tar.gz` extension? Or should we save files with `tar.gz` and `zip` extensions?

File: .github/workflows/publish.yml (hunk 2 above)
Comment (dmvict) [reply .github/workflows/publish.yml:165]: Hello @kevaundray 

Do we save the `nightly` prefix? Should I change it to `preview`?

Comment (kevaundray) [reply .github/workflows/publish.yml:165]: Hello, lets keep nightly prefix

File: .github/workflows/publish.yml (hunk 1 above)
Comment (kevaundray) [reply .github/workflows/publish.yml:162]: This PR should not change anything to do with tar.gz, since we are just adding nightly tags

**Comment by dmvict** [comment]
Hello @phated 

I fixed the previous issues a time ago. Now I merged the new changes. 

Please, review the changes and approve/add new comments.

**Review by kevaundray — APPROVED** [review]

**Pull request merged**